packages/utils/src/Defer.ts

```diff
@@ -0,0 +1,68 @@
+import { once } from 'lodash'
+
+type ResolveFn<T> = (value: T | PromiseLike<T>) => void
+
+type RejectFn = (reason?: Error) => void
+
+const noopExecutor = () => {}
+
+/**
+ * Deferred promise allowing external control of resolve / reject.
+ */
+export class Defer<T> extends Promise<T> {
+    private readonly resolveFn: ResolveFn<T>
+    private readonly rejectFn: RejectFn
+    private readonly ensureNoopCatchAttached: () => void
+    private settled = false
+
+    constructor(executor: (resolve: ResolveFn<T>, reject: RejectFn) => void = noopExecutor) {
+        let localResolve: ResolveFn<T> | undefined
+        let localReject: RejectFn | undefined
+        super((resolve, reject) => {
+            localResolve = resolve
+            localReject = reject
+            executor(resolve, reject)
+        })
+        if (localResolve === undefined) {
+            throw new Error('invariant violation: resolveFn was undefined')
+        }
+        if (localReject === undefined) {
+            throw new Error('invariant violation: rejectFn was undefined')
+        }
+        this.resolveFn = localResolve
+        this.rejectFn = localReject
+        this.ensureNoopCatchAttached = once(() => {
+            super.catch(() => {})
+        })
```

Comment:
I couldn't figure out a way to get `this.catch(() => {}` or `super.catch(() => {})` to work directly in constructor without leading to `Maximum call stack size exceeded` error.